include/seqan3/argument_parser/validators.hpp

```diff
@@ -297,68 +290,347 @@ class file_ext_validator
+ * \ingroup argument_parser
+ * \implements seqan3::validator_concept
+ *
+ * \details
+ *
+ * On construction, the validator can receive a list (std::vector over std::string) of valid file extensions.
+ * The struct acts as a functor that throws a seqan3::parser_invalid_argument exception whenever a given filename's
+ * extension (sts::filesystem::path) is not in the given list of valid file extensions, if the file does not exist, or
+ * if the file does not have the proper read permissions.
+ *
+ * \snippet test/snippet/argument_parser/validators_input_file.cpp usage
+ *
+ * \note The validator works on every type that can be implicitly converted to std::filesystem::path.
+ */
+class input_file_validator : public file_validator_base
+{
+public:
+    //!\brief Type of values that are tested by the validator.
+    using file_validator_base::value_type;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    input_file_validator() = default;                                         //!< Defaulted.
+    input_file_validator(input_file_validator const &) = default;             //!< Defaulted.
+    input_file_validator(input_file_validator &&) = default;                  //!< Defaulted.
+    input_file_validator & operator=(input_file_validator const &) = default; //!< Defaulted.
+    input_file_validator & operator=(input_file_validator &&) = default;      //!< Defaulted.
+    ~input_file_validator() = default;                                        //!< Defaulted.
+
+    //!\brief Import constructor for setting valid extensions.
+    using file_validator_base::file_validator_base;
+    //!\}
+
+    //!\brief Import function-call operator.
+    using file_validator_base::operator();
+
+    /*!\brief Tests whether path is an existing regular file and is readable.
+     * \param file The input value to check.
+     * \throws parser_invalid_argument if the validation process failed or std::filesystem::filesystem_error if some
+     *         unexpected errors during the filesystem operations occur.
+     */
+    void operator()(std::filesystem::path const & file) const override
+    {
+        if (!std::filesystem::exists(file))
+            throw parser_invalid_argument(detail::to_string("The file ", file, " does not exist!"));
+
+        // Check if file has any read permissions.
+        if (!path_is_readable(file))
+            throw parser_invalid_argument(detail::to_string("Cannot read ", file, "!"));
+
+        // Check extension.
+        file_validator_base::validate_path_impl(file);
     }
 
     //!\brief Returns a message that can be appended to the (positional) options help page info.
     std::string get_help_page_message() const
     {
-        return detail::to_string("File name extension must be one of ", std::view::all(extensions), ".");
+        return detail::to_string("Valid input file formats: ",
+                                 file_validator_base::extensions | std::view::join(std::string{", "}),
```

Comment:
If do not use `join`, the list is displayed as `Valid input file formats: [fasta, fa]` which is sufficient isn't it?

Reply:
I check it again. I am pretty sure that I added it because the output was strangely without the space.